File: src/test/java/micronaut/playground/web/HelloControllerTest.java
package micronaut.playground.web;

import io.micronaut.test.annotation.MicronautTest;
import io.micronaut.test.annotation.MockBean;
import io.reactivex.Single;
import micronaut.playground.client.Hello;
import micronaut.playground.client.HelloClient;
import micronaut.playground.domain.Greeter;
import micronaut.playground.domain.GreeterImpl;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import javax.inject.Inject;

import java.time.LocalDateTime;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

@MicronautTest
class HelloControllerTest {

  @Inject
  private Greeter greeter;

  @Inject
  private HelloClient client;

  @AfterEach
  void tearDown() {
    verifyNoMoreInteractions(greeter);
  }

  @Test
  void hello() {
    when(greeter.greetHello("gukguk"))
        .thenReturn(Single.just(new Hello("Test", "gukguk", 1L, LocalDateTime.now())));

    Hello response = client.hello("gukguk").blockingGet();
    assertEquals("Test", response.getGreeting());
    assertEquals("gukguk", response.getName());

    verify(greeter).greetHello("gukguk");
  }

  @MockBean(GreeterImpl.class)
  Greeter greeter() {
    return mock(Greeter.class);
  }
}

File: src/test/kotlin/micronaut/playground/web/HelloControllerTest.kt
package micronaut.playground.web

import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.micronaut.test.extensions.kotest5.annotation.MicronautTest
import micronaut.playground.client.HelloClient

@MicronautTest(transactional = false)
class HelloControllerTest(
    private val client: HelloClient
) : BehaviorSpec({

    given("the hello endpoint") {
        `when`("the greeter is called with Zay") {
            val result = client.hello("Zay").block()
            then("the greeting is Test") {
                result.greeting shouldBe "Test"
            }
            then("the name is Zay") {
                result.name shouldBe "Zay"
            }
            then("the count is 0") {
                result.count shouldBe 0
            }
        }
    }
})